Add subject and pre-header to email editor sidebar Settings panel

lysyjan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

- This PR adds new rich-text components to the email editor sidebar, where it is possible to configure the subject and preview text. 

Closes WOOPLUG-3680.

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|<img width="1277" alt="Screenshot 2025-04-16 at 17 11 04" src="https://github.com/user-attachments/assets/65389f09-bbc4-4f83-9376-57ba4f3303ef" />|<img width="1278" alt="Screenshot 2025-04-16 at 16 50 28" src="https://github.com/user-attachments/assets/63b775e3-3972-4fca-9a5e-e00a201eb83a" />|



<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable the block-based email editor from WooCommerce → Settings → Advanced → Features → Block Email Editor (alpha)
2. Open an email in the editor WooCommerce → Settings → Emails
3. Select an email and set the subject and preheader, and use the personalization tags there
4. Replicate sending for the selected email (because preview does not work now)
5. Check that the received email has the configured subject and preheader correctly

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

lysyjan committed:
Rename DetailsPanel to SettingsPanel and change title

lysyjan committed:
Move template selection in email editor into settings panel

lysyjan committed:
Fix CSS className for the email editor settings panel

lysyjan committed:
Align the button opening Personalization tags modal to right

lysyjan committed:
Use richText component supporting Personalization Tags for subject and preheader

lysyjan committed:
Register rest field for passing subject and heading

lysyjan committed:
Update subject and heading in the option table

lysyjan committed:
Use default subject and heading as a placeholder

lysyjan committed:
Refactor preheading into preview text

lysyjan committed:
Add subject personalization into WC transactional emails

lysyjan committed:
Enhance email subject handling for customer_refunded_order type by adding separate fields for full and partial refund subjects in the email editor

lysyjan committed:
Add get_preheader method to WC_Email class and update BlockEmailRenderer to utilize it for rendering preheader text in block emails

lysyjan committed:
Add changelogs

lysyjan committed:
Fix BlockEmailRendererTest and set default value in class-wc-email for preheader to empty string

costasovo commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/email-editor-integration/sidebar_settings.tsx:
```diff
@@ -0,0 +1,59 @@
+ */
+import { NAME_SPACE } from './constants';
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
```
Could you please replace this with @ts-expect-error? This will give us a notification when the error is fixed e.g., via updating a package, etc. Also, please provide a basic comment on why we are accepting the error.

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/EmailApiController.php:
```diff
@@ -24,9 +37,13 @@ public function get_email_data( $post_data ): array {
 			'subject' => $post_option['subject'] ?? null,
 			'heading' => $post_option['heading'] ?? null,
+			'default_subject' => $email->get_default_subject(),
+			'default_heading' => $email->get_default_heading(),
```
Why do we need to pass these defaults as separate properties? Can't we just use them as a subject fallback directly?

```php
'subject' => $post_option['subject'] ?? $email->get_default_subject(),
```

lysyjan replied:
@costasovo The default values are used as placeholders, not as default values. Do you think we should use another name in the response here?

costasovo replied:
@lysyjan Placeholders, I see. How does it work during sending, when a user doesn't save any Subject and just keeps the placeholder? What subject would be used?

lysyjan replied:
Sending uses the [WC_Email::get_subject()](https://github.com/woocommerce/woocommerce/blob/b9956f17234492c7272aed5f16684d0ca2062aa3/plugins/woocommerce/includes/emails/class-wc-email.php#L485-L505) method, which handles logic such as the use of the default subject if needed. I'd say the logic is the same as the old email setup page.

costasovo commented on plugins/woocommerce/includes/emails/class-wc-email-cancelled-order.php:
```diff
@@ -97,6 +97,10 @@ public function trigger( $order_id, $order = false ) {
 				$this->placeholders['{order_billing_full_name}'] = $this->object->get_formatted_billing_full_name();
 			}
 
+			if ( $this->block_email_editor_enabled ) {
+				$this->personalizer->set_context( $this->prepare_context_data( $this->personalizer->get_context() ) );
+			}
+
```
I think we should try to avoid modifying individual emails as much as possible. In this case we repeat the same two lines in every email type class. This will add a requirement that every plugin that adds email type will need to add these two lines.

Could we move it to `WC_Email::send` ? It receives the subject, so we could personalize it directly in the method or add a new filter there. I think it would be better to avoid changing every sinle WC_Email. What do you think?

lysyjan replied:
> Could we move it to WC_Email::send ? It receives the subject, so we could personalize it directly in the method or add a new filter there. I think it would be better to avoid changing every sinle WC_Email. What do you think?

I am not sure about this approach. Because the replacement would be used only in the case we send the email through this method. But what about a preview? For example, sending a preview is managed in the PHP email editor package through `wp_mail`. Another use case can be HTML preview.  I prefer keeping the setting context for personalization tags in the same place as it was for the old placeholders. Maybe it would be better to chat about it if you want.

costasovo replied:
I see that you've implemented a different approach and now we don't have to touch every single email 👍

costasovo commented on plugins/woocommerce/includes/emails/class-wc-email.php:
```diff
@@ -1429,6 +1451,29 @@ protected function get_option_or_transient( string $key, $empty_value = null ) {
+	 * @param array $context Previous version of context data.
+	 * @return array Context data for personalization
+	 */
+	protected function prepare_context_data( array $context ): array {
+		$context['recipient_email'] = $this->get_recipient();
+		$context['order']           = $this->object instanceof WC_Order ? $this->object : null;
+		// For emails of type new_user or reset_password we want to set user directly from the object.
+		if ( $this->object instanceof \WP_User ) {
+			$context['wp_user'] = $this->object;
+		} elseif ( $this->object instanceof WC_Order ) {
+			$context['wp_user'] = $this->object->get_user();
+		} else {
+			$context['wp_user'] = null;
+		}
+		$context['wc_email'] = $this;
+		return $context;
+	}
```
I see that we already have this logic in 
https://github.com/woocommerce/woocommerce/blob/5e8830758700e6c5e3655adcd2e571ab8b945ab0/plugins/woocommerce/src/Internal/EmailEditor/BlockEmailRenderer.php#L134-L141

Can we somehow avoid duplicating the logic for creating the context? I think that in the future when we will implement the Woo Order blocks we may need to be able to pass such a context directly into block renderers. So having a service that could build or build and carry sending context would be helpful.

A brief draft I have in my mind could look like this:
```php
class SendContext {
  private $context = [];
  
  public function addContextFromWCEmail( WC_Email $email) {
     ...the code you have here
  }
  
  public function getContext() {
    return $this->context;
  }
}
```
Do you think something like that could work?

lysyjan replied:
I was looking at this, and we can remove the setting context in the `BlockEmailRenderer` class because using this class and the main method is a part of the `get_content()` method. In my opinion, the context should be set before using `get_content()`.

costasovo replied:
Makes sense. It is good that we have that in one place. I think that in the future, we might need context in different places than in the personalizer (e.g., blocks that render an order table might need it as well), but we can address that later when we have more details.

costasovo reviewed (commented):
@lysyjan Great job!

I found a few places I would like to discuss. Please see my other comments.
Note: I didn't test the feature locally in detail, because I would like to discuss the comments first.

lysyjan committed:
Use @ts-expect-error instead of @ts-ignore

lysyjan committed:
Remove redundant context configuration

lysyjan committed:
Refactor setting context to avoid repetition

I added a method used when we want to personalize content to ensure the context is already set.

lysyjan commented:
@costasovo I refactored the code related to setting context for personalization to reduce code repetition. There are three new commits. Please let me know what you think about this approach.

costasovo commented on plugins/woocommerce/includes/emails/class-wc-email.php:
```diff
@@ -479,12 +486,37 @@ public function get_additional_content() {
+	 * @return void
+	 */
+	protected function ensure_personalization_context_is_set(): void {
+		if ( $this->context_set ) {
+			// Context is already set, skip reconfiguration.
+			return;
+		}
```
I think there might be an issue with this caching. Can we ensure we don't cache "empty" context?

The context data is set in trigger methods in classes extending WC_Email. So I have a concern that when somebody calls
`$email->getSubject()` before the trigger method is called, it may cause us to store an empty context, and we won't update it after the trigger is actually called.

costasovo reviewed (commented):
@lysyjan Thank you for the improvements. I like how the personalization is aligned with the original placeholder replacements.

I added one comment with a question about caching the context.

I did some basic testing locally and set subject and preheder for a couple of emails and email I received had proper subject and preheader.

I noticed a small issue with the subject, though. It seems that because we use Richtext it may contain some HTML.

**Here is how I set it:**
![Screenshot 2025-04-29 at 13 33 54](https://github.com/user-attachments/assets/afd0c43d-35e1-4861-9bdc-47818df50e06)

**Here is what I received:**
Notice `<br>` tag in the subject.
![Screenshot 2025-04-29 at 13 32 55](https://github.com/user-attachments/assets/3bf56762-8494-4d02-a64b-5fbd2a4dda6a)
After I inserted the second personalization tag to subject via modal it added an extra line between the fris couple of words in the subject and newly inserted tag. Perhaps because I placed the cursor there. I'm not sure. It is probably not that important. The problem is that new lines were interpreted as <br> tags. Tags are not supported in subject in email clients so I think we should strip them.

lysyjan committed:
Update context setting logic in class-wc-email

lysyjan committed:
Sanitize email subject by stripping all HTML tags in class-wc-email

lysyjan commented:
@costasovo Thank you for your ideas and testing. I added two more commits with tiny adjustments. I think it solves the mentioned issues.

lysyjan committed:
Use WCTransactionalEmailPostsManager instead of the constant from Integration class

lysyjan committed:
Create a new class for easier personalization and better reusability

costasovo reviewed (approved):
@lysyjan Great job!

Kudos for going with composition for the TransactionalEmailPersonalizer 👍. Great idea!

### Testing performed

I tested with the "Processing order" email which is sent to a shopper after they make an order.

I filled custom subject and preheader with personalization tags:

<img width="299" alt="Screenshot 2025-05-02 at 12 47 04" src="https://github.com/user-attachments/assets/68e3e1de-179c-4439-a6cc-996e865469d0" />

and I can see I received the email with proper values in subject and preheader.

<img width="452" alt="Screenshot 2025-05-02 at 12 46 21" src="https://github.com/user-attachments/assets/10877f4e-7d81-4204-bbce-314820f01e75" />

I also tested with an empty subject and I see it uses the default subject:

<img width="486" alt="Screenshot 2025-05-02 at 12 47 40" src="https://github.com/user-attachments/assets/24aa084e-2322-4e06-bd8e-8530720241de" />
